### ISSUE TITLE
Ops with several optional inputs

### ISSUE THREAD

**Issue by dzhulgakov** [onnx/onnx#151]
Some of the operators (e.g. OptimizedRNN, see #149) require several inputs that can be independently present or absent.

Possible solution: allow some of the positional inputs to be 'None' in the graph with inputs represented as empty string in the proto. Schema checker would enforce that they can appear only after the MinArgs provided.

If not objections - I can cut the diff.

**Comment by ezyang** [comment]
As an extra data point, in PyTorch we model optional inputs using an `Undefined` operator, which produces an "undefined tensor", much like null.

On the other hand, NULL was Tony Hoare's billion dollar mistake ;)

**Comment by bddppq** [comment]
>  Schema checker would enforce that they can appear only after the MinArgs provided.
Why adding this requirement?

**Comment by ezyang** [comment]
It's better to distinguish between "takes a tensor" and "takes a tensor or NULL" and reject NULL where it is not accepted (which is most places.)

**Comment by yuanbyu** [comment]
Yes it is probably needed for the kind of flexibility you would like to have. Should we enforce at the IR graph level that if an input is optional it must take a real tensor or `None`? Also, I am fine with empty string, but another option is to introduce a unique name `None` or `Default`. Any changes to the op schema?

**Comment by ezyang** [comment]
Using a name like None/Default is a BC-breaking change. I also feel like it makes exporter logic modestly more complicated (because it has to test for reserved keywords), so I'd prefer not to do it.

**Comment by dzhulgakov** [comment]
I'd go with empty string as it's easier to handle.

To make op schema more sane we should probably add "optional" argument to Input() that overrides behavior of NumInputs.

But the very bare minimum is that we don't change anything and implicitly allow optional args if # of inputs is between MinInputs and MaxInputs.

**Comment by yuanbyu** [comment]
Another possible (and more general) solution to this problem is something I suggested a while ago. If we add the following field to NodeProto, we can handle multiple operator arguments that can independently take 0, 1, or more real inputs.

   ```
  // The number of inputs for each argument of the operator takes
  // A formal parameter of the op may take a variable number of inputs
  // that is only known when this node is constructed.
  repeated int32 input_arg_count;
```

**Comment by gramalingam** [comment]
Empty string sounds fine to me.

As Yuan says, this is also a special case of the more general problem of supporting input arguments that may be a list of tensors. So, it is worth thinking about the solution to the more general problem also here. Do we want to support two different mechanisms (both input_arg_count and empty strings)?

I would also like to understand the role of MinInputs. Is the idea that none of the first MinInputs can be optional? It seems like a good idea to specify optional-arguments in the op-schema (and then "MinInputs" is just a short-hand notation).

I assume that trailing missing-inputs can be completely omitted, and that empty-string is required only when some subsequent input is present?

**Comment by gramalingam** [comment]
Another possibility is to enrich the language of "expressions" a little bit. For example, a tensor-expression can be a (variable) reference to a name (string) that is bound in a different statement, which is currently the only kind of tensor-expression supported. But we have other needs as well:

(a) Representing a special value (Null/None) to denote the absence of a tensor,
(b) A literal (constant-valued tensor), which is currently supported separately via initializers,
(c) A list of tensors (which could be supported by input_arg_count).

**Comment by dzhulgakov** [comment]
I think we should separate lists of inputs and optional inputs - they are semantically different even though optional input can be represented as list of 0..1.

For lists it's basically tradeoff between making graph structure more complicated and user-friendliness. Unless op really needs 2 separate lists of tensors I don't think it's strictly necessary to have lists support (as we can always put the list at the very end of the op).

Yuan's suggestion of `input_arg_count` is nice but it might be a bit hard to parse if we use it too much for optional arguments. Then it'd be mentally hard to parse the arguments order - empty string or other placeholder seem a bit easier.

So how about going with empty string for now, adding some annotations for optionality and with lists waiting to see whether we'd run into problems too much or not

**Comment by yuanbyu** [comment]
Empty string is fine with me. 

Once we have good support for optional inputs, do you think we should revisit the weight packing of OptimizedRNN? Perhaps we could use separate inputs for W, B, R, ...?

**Comment by linkerzhang** [comment]
I'm sorry that I didn't add my personal opinions when this issue open. 

Given now we have both "optional" input in between input list and also variadic input (normally should be the last input of an op). I think using input_arg_count that Yuan suggested is a better way to unify both cases. Otherwise, I believe in any runtime implementation, we need to handle these two cases separately. Make sense please?

**Comment by dzhulgakov** [comment]
My worry with input_arg_count is that we have to add it to every Node (so that the input is parseable without having schema) even though it's necessary only for a small subset of nodes. Empty string is less intrusive in this sense.

**Comment by linkerzhang** [comment]
Aha, I personally think the overhead of adding it to nodes is ok to take. It makes things more clearer and consistent. That means we will be able to have an array like "1, 0, 0, 1, 3" in a node to specify its number of inputs of each formal parameter.  With current design, we'll have empty strings to represent 0 arg for one input. With the last variadic input, we'll anyway have to go thru the whole thing to identify the number of args for last variadic input. Two piece of logic will be needed.

Stepping back, if we don't want to take the overhead, we can set a default value of this array by default, say, [1, 1, 1, 1, 1, ... 1,] (all are 1), and we don't store the default array in nodes. 

Thoughts?

**Comment by gramalingam** [comment]
(a) If we have input_arg_count, it makes sense to make its default value (if missing) to be all ones.

(b) The tradeoff here is between more flexibility (with multiple ways of representing a missing parameter) and minimality in the IR. I think this tradeoff arises in multiple situations, so if we can agree on the general principle we want to use, it will help. Flexibility may simplify the job of tools that produce a model, while minimality will simplify the job of the consumers of models. I prefer giving more weight to yet another factor, which is: "simplicity" of the IR (easier to understand), which does overlap somewhat with "minimality".

**Comment by linkerzhang** [comment]
Thank you @gramalingam for the great comments! So,  what's your preference please? I mean whether to use "input_arg_count" solution. In one way, I'm seeing using it is a little bit more consistent way to handle both "optional" and "variadic" inputs. In another way, I'm not sure whether we'll have [1, 1, 3, 0, 2, 1, 3] case later, although normal programming language does not support this :).

**Comment by gramalingam** [comment]
Originally, OptimizedRNN was a motivation for optional inputs. With the splitting of OptimizedRNN into multiple separate ops, do we still have a need for optional inputs? Do we currently have a need for input_arg_count? I am just trying to understand if this discussion is for future extension or we need it right now.

One concern I have is the following: even though the ONNX standard is supposed to be a specification for only the serialized representation of the model, the design choice here seems to be partly motivated by the underlying (current) implementation of operators, especially in terms of how input parameters are passed in to the operator implementation. 

Since ONNX-ML also makes use of the richer "sequence" type (for input values), would we then support both? The richer sequence can have length that is dynamically determined and is represented as a single input of sequence type; while the above-discussed one has a length that is statically known and is represented as a "variadic" input.  Do we need to support both?

**Comment by dzhulgakov** [comment]
OptimizedRNN still needs it as e.g. LSTM has optional initial states and optional sequence lengths - both are optional independently.

The current specification is very simple, i.e. positional arguments to operators. I don't think it depends much on how implementations handle it.

As for "sequence" type - it's fundamentally different. Here we're talking about passing several other tensors from the graph together as an op, not representing a single value of type list. Equivalent would be having Pack operator which would take (still variadic) number of input arguments and produce sequence as an output. But it introduces additional indirection in the graph and makes it harder.

The reason I don't like input_arg_count is that it changes the default behavior in unexpected way. If I have optional Inputs 2 - X and Input 3 - Y with input_arg_count I don't know any more whether op.input[2] represents X or Y. Since it's not a common path, it's more probably that people would misuse it. Whereas with empty strings it's kind of obvious that the input is missing if op.input[2] is empty. And if implementation doesn't handle it properly, it'd fail in quite predictable way by tripping over an empty string.
